### ISSUE TITLE
Error while installing bindings

### ISSUE THREAD

**Issue by joaomhmpereira** [bitwuzla/ocaml-bitwuzla#8]
Hi, I'm having problems installing the bindings. I've tried installing through Opam and building from source but always get the same error. I'm using macOS v15.0 (Sequoia).

Error:

```shell
File "vendor/dune", line 52, characters 11-19:
52 |     (names bitwuzla)
                ^^^^^^^^
In file included from bitwuzla.cpp:18:
In file included from /opt/homebrew/include/node/node.h:73:
In file included from /opt/homebrew/include/node/v8.h:23:
/opt/homebrew/include/node/cppgc/common.h:8:10: fatal error: 'v8config.h' file not found
#include "v8config.h"  // NOLINT(build/include_directory)
         ^~~~~~~~~~~~
1 error generated.
```

Looks like it's including things from Node.js instead of the `node` from the solver. Any help?

**Comment by recoules** [comment]
Dear @joaomhmpereira,

Sorry for this issue. I guess it is because the *include directive* (`-I`) of `homebrew/include` is before the `include` of the project in the `dune` file.
Could you tell me if the following patch fixes your issue ?

```diff
diff --git a/vendor/bitwuzla b/vendor/bitwuzla
--- a/vendor/bitwuzla
+++ b/vendor/bitwuzla
@@ -1 +1 @@
-Subproject commit 31e0296ebf5685a5effd4accaffe2e8dc98a0b5b
+Subproject commit 31e0296ebf5685a5effd4accaffe2e8dc98a0b5b-dirty
diff --git a/vendor/dune b/vendor/dune
index 66a9d5faff8..0ae93472f1c 100644
--- a/vendor/dune
+++ b/vendor/dune
@@ -15,7 +15,7 @@
    (archive_name bitwuzla-utils)
    (language cxx)
    (names hash_pair logger printer resources statistics)
-   (flags :standard -std=c++17 -I. -I.. -DNDEBUG)
+   (flags -I. -I.. :standard -std=c++17 -DNDEBUG)
    (extra_deps ../../../.bitwuzla_tree)))
  (subdir
   src/lib
@@ -38,7 +38,7 @@
     wheel_factorizer
     rng)
    (include_dirs bitblast bv ls rng)
-   (flags :standard -std=c++17 -I. -I.. -DNDEBUG)
+   (flags -I. -I.. :standard -std=c++17 -DNDEBUG)
    (extra_deps ../../../.bitwuzla_tree)))
  (subdir
   src/api
@@ -51,11 +51,11 @@
     (language cxx)
     (names bitwuzla)
     (flags
-     :standard
-     -std=c++17
      -I../..
      -I../../lib
      -I../../../include
+     :standard
+     -std=c++17
      -DNDEBUG)
     (extra_deps ../../config.h ../../../../.bitwuzla_tree))))
  (subdir
@@ -69,7 +69,7 @@
      (archive_name bitwuzla-smt2)
      (language cxx)
      (names lexer)
-     (flags :standard -std=c++17 -I../.. -DNDEBUG)
+     (flags -I../.. :standard -std=c++17 -DNDEBUG)
      (extra_deps ../../../../.bitwuzla_tree))))
   (rule
    (deps ../COPYING)
@@ -174,7 +174,7 @@
    type_data
    type_manager)
   (include_dirs include src src/lib ../symfpu ../cadical/src)
-  (flags :standard -std=c++17 -Isrc/lib/ -I../cadical/src -I.. -DNDEBUG)
+  (flags -Isrc/lib/ -I../cadical/src -I.. :standard -std=c++17  -DNDEBUG)
   (extra_deps ../.bitwuzla_tree)))
 
 (subdir
```

Regards,
Frédéric

**Comment by joaomhmpereira** [comment]
Thanks for the swift response! 
Unfortunately, the patch did not fix the problem. As a temporary solution, I uninstalled Node.js, but now I'm getting another error:

```shell
File "dune", line 17, characters 9-18: 
17 |   (names cxx_utils cxx_stubs)
              ^^^^^^^^^
cxx_utils.cpp:33:43: error: no member named 'to_zarith' in 'bzla::BitVector'
  return d_node->value<bzla::BitVector>().to_zarith();
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
cxx_utils.cpp:37:18: error: out-of-line definition of 'to_zarith' does not match any declaration in 'bzla::BitVector'
bzla::BitVector::to_zarith() const
                 ^~~~~~~~~
2 errors generated.
File "dune", lines 34-40, characters 0-110:      
34 | (rule
35 |  (deps api/options.exe)
36 |  (target options.ml)
37 |  (action
38 |   (with-stdout-to
39 |    %{target}
40 |    (run %{deps}))))
Command got signal SEGV.
```

**Comment by recoules** [comment]
For this one, I think a `dune clean` should fix it.

**Comment by joaomhmpereira** [comment]
It solves the part regarding `zarith`, but I still get:
```shell
❯ dune build @install
File "dune", lines 34-40, characters 0-110:
34 | (rule
35 |  (deps api/options.exe)
36 |  (target options.ml)
37 |  (action
38 |   (with-stdout-to
39 |    %{target}
40 |    (run %{deps}))))
Command got signal SEGV.
```

**Comment by recoules** [comment]
Well, I am sorry, I have no clue for now.
I will have to make some test to figure out what is happening, but I am not sure if I have access to a macOS v15.

Just to be sure, did you try to build and run the "*vanilla*" `bitwuzla` (https://github.com/bitwuzla/bitwuzla) ?
If it works, then, the issue comes from the OCaml wrapper, but I there are some error, maybe we will need to patch the original source code.

**Comment by recoules** [comment]
Ok, I managed to get a macOS v15 working station.

I was able to reproduce the original bug (`nodejs`) but not the segmentation fault.

The following patch fixed the first issue for me (almost the same as yesterday but changes two files instead of one).

```diff
diff --git a/dune b/dune
index 6b904d0..c58eab3 100644
--- a/dune
+++ b/dune
@@ -20,7 +20,7 @@
    vendor/bitwuzla/src
    vendor/bitwuzla/src/lib
    (lib zarith))
-  (flags :standard -std=c++17 -I/opt/homebrew/include))
+  (flags :standard -std=c++17 -Ivendor/bitwuzla/src -I/opt/homebrew/include))
  (foreign_archives
   vendor/bitwuzla/src/api/cpp/bitwuzla-api
   vendor/bitwuzla/bitwuzla-cxx
diff --git a/vendor/dune b/vendor/dune
index 66a9d5f..8b1742e 100644
--- a/vendor/dune
+++ b/vendor/dune
@@ -51,11 +51,11 @@
     (language cxx)
     (names bitwuzla)
     (flags
-     :standard
      -std=c++17
      -I../..
      -I../../lib
      -I../../../include
+     :standard
      -DNDEBUG)
     (extra_deps ../../config.h ../../../../.bitwuzla_tree))))
  (subdir
@@ -174,7 +174,7 @@
    type_data
    type_manager)
   (include_dirs include src src/lib ../symfpu ../cadical/src)
-  (flags :standard -std=c++17 -Isrc/lib/ -I../cadical/src -I.. -DNDEBUG)
+  (flags  -std=c++17 -Isrc/lib/ -I../cadical/src -I.. -Isrc :standard -DNDEBUG)
   (extra_deps ../.bitwuzla_tree)))
 
 (subdir
```
Could you tell me if it works for you ?

Note : is the segfault also happens with an Opam installation after you uninstalled `nodejs` ?

**Comment by recoules** [comment]
Hi @joaomhmpereira, did you manage to make it works ?

**Comment by joaomhmpereira** [comment]
Sorry for the delay.  Yes, the patch does seem to fix the `nodejs` issue!

> does the segfault also happen to an Opam installation after you uninstalled `nodejs` ?

Yes, it also occurs but I'm confused if it is related with the bindings or not. I get this error:

```shell
[ERROR] The compilation of bitwuzla-cxx.0.5.0 failed at "dune build -p bitwuzla-cxx -j 7 @install".

#=== ERROR while compiling bitwuzla-cxx.0.5.0 =================================#
# context     2.1.5 | macos/arm64 | ocaml-base-compiler.5.1.0 | https://opam.ocaml.org#fe8dffac61bbab376897c0f4e7cc5b7b8c4eecc1
# path        ~/.opam/5.1.0/.opam-switch/build/bitwuzla-cxx.0.5.0
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p bitwuzla-cxx -j 7 @install
# exit-code   1
# env-file    ~/.opam/log/bitwuzla-cxx-5501-639d41.env
# output-file ~/.opam/log/bitwuzla-cxx-5501-639d41.out
### output ###
# [...]
# In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX15.0.sdk/usr/include/c++/v1/__atomic/atomic.h:12:
# In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX15.0.sdk/usr/include/c++/v1/__atomic/atomic_base.h:12:
# In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX15.0.sdk/usr/include/c++/v1/__atomic/atomic_sync.h:12:
# In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX15.0.sdk/usr/include/c++/v1/__atomic/contention_t.h:12:
# /Library/Developer/CommandLineTools/SDKs/MacOSX15.0.sdk/usr/include/c++/v1/__atomic/cxx_atomic_impl.h:461:59: error: reference to unresolved using declaration
# __cxx_atomic_fetch_add(__cxx_atomic_base_impl<_Tp*>* __a, ptrdiff_t __delta, memory_order __order) _NOEXCEPT {
#                                                           ^
# /Library/Developer/CommandLineTools/SDKs/MacOSX15.0.sdk/usr/include/c++/v1/cstddef:60:1: note: using declaration annotated with 'using_if_exists' here
# using ::ptrdiff_t _LIBCPP_USING_IF_EXISTS;
# ^
# fatal error: too many errors emitted, stopping now [-ferror-limit=]
# 20 errors generated.
```

**Comment by recoules** [comment]
Hum...

Could you send me the full content of `~/.opam/log/bitwuzla-cxx-5501-639d41.out` and `~/.opam/log/bitwuzla-cxx-5501-639d41.env` ?

By the way, do you still get a segmentation fault when building from source with `dune` ?

**Comment by joaomhmpereira** [comment]
After some digging, I found several other people complaining about a similar problem when upgrading the macOS version (for instance here: https://apple.stackexchange.com/questions/458062/unable-to-find-include-file-on-newly-updated-macosc-c/459037#459037). Since I recently updated to v15.0 I think that the error might be related with that and not with this package.

Thanks for all the help in this process! I'm happy with closing this issue if you are
